sklearn/utils/validation.py

```diff
@@ -397,7 +397,7 @@ def _num_samples(x):
     if hasattr(x, "shape") and x.shape is not None:
         if len(x.shape) == 0:
             raise TypeError(
-                "Singleton array %r cannot be considered a valid collection." % x
+                f"Expected array-like or collection type, got type `{type(x)}` instead."
```

Comment:
I think I would mostly replace "Singleton array" by "Numpy scalar" in the original error message (or maybe only "scalar" if we think that mentioning numpy can be confusing if we are passing non-numpy types like torch scalar)

Reply:
and/or mention more explicitly "input `x` was a scalar i.e. `len(x.shape) == 0` you should instead pass an input `x` with `len(x.shape) > 0`

Reply:
Actually Numpy uses "array scalar" in their [doc](https://numpy.org/doc/stable/reference/arrays.scalars.html so I would say) so maybe something like this. Better suggestion more than welcome!

```suggestion
                f"Input 'x' should have at least 1 dimension i.e. satisfy len(x.shape) > 0, got array scalar {!r} instead."
```

Reply:
Yes, I like it. I just made "array scalar" into "scalar" just to make it the right message for number types as well.